### PR TITLE
Feat/added toml supersim config closes #104

### LINKED ISSUE #104: Config Improvements
- toml file that gets parsed and the values are passed in as cli args

### PR DESCRIPTION
**Pull request by s29papi**
**Description**

This pull request implements significant improvements to the configuration handling of the application. The changes include the addition of a TOML configuration file that is parsed, allowing values to be passed in as command-line arguments. The configuration values from the TOML file take precedence over those provided via CLI flags, ensuring a more flexible and user-friendly configuration experience.

**Tests**

Comprehensive tests have been added to verify the correct parsing of the TOML file and the proper application of configuration values. The tests ensure that:
- The application correctly reads and applies values from the TOML file.
- CLI arguments are overridden by values from the TOML file when both are provided.
- All edge cases, such as missing or default values, are handled appropriately.

**Additional context**

These improvements address the need for a more robust configuration management system, allowing users to define their settings in a structured format while still providing the flexibility of command-line overrides. This change enhances usability and aligns with best practices for configuration management.

**Metadata**

- Fixes #104

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -31,7 +31,44 @@ brew install ethereum-optimism/tap/supersim
 
 Download the executable for your platform from the [GitHub releases page](https://github.com/ethereum-optimism/supersim/releases).
 
-### 3. Start `supersim` in vanilla mode
+### 3. Configure `supersim` (Optional)
+
```
Comment (jakim929) [inline README.md:35]: this readme section is great but can we move it down to a separate section `## Configuration` after `## First Steps`. The Getting Started should stay as short as possible

File: config/cli.go [hunk 2]
```diff
@@ -154,32 +156,40 @@ type CLIConfig struct {
...
 
-		LogsDirectory: ctx.String(LogsDirectoryFlagName),
+	L1Port         uint64 `toml:"l1_port"`
+	L2StartingPort uint64 `toml:"l2_starting_port"`
 
-		L1Host: ctx.String(L1HostFlagName),
-		L2Host: ctx.String(L2HostFlagName),
-	}
+	InteropAutoRelay bool   `toml:"interop_autorelay"`
+	InteropDelay     uint64 `toml:"interop_delay"`
```
Comment (jakim929) [inline config/cli.go:175]: TOMLConfig and ForkTOMLConfig should be combined with the ForkCLIConfig and CLIConfig. I think we can just add struct tags to the CLIConfig and ForkCLIConfig.

File: .gitignore [hunk 3]
```diff
@@ -30,3 +30,7 @@ go.work.sum
 /main
 dist/
 cache/
+
+# Ignore customized configuration files
+*.toml
+!*.toml.template
```
Comment (jakim929) [inline .gitignore:36]: can probably remove this - People primarily consume this package as a binary so it's rare that people will directly try to commit their toml to the repo

File: config/cli.go [hunk 4]
```diff
@@ -244,3 +254,87 @@ func validateHost(host string) error {
 
 	return nil
 }
+
+func readTOMLConfig(cfg *CLIConfig) error {
```
Comment (jakim929) [inline config/cli.go:258]: nit: maybe rename to "applyTOMLConfig" since this does more than reading

File: config/cli.go [hunk 5]
```diff
@@ -244,3 +254,87 @@ func validateHost(host string) error {
...
+			}
+		}
+	} else if !os.IsNotExist(err) {
+		return fmt.Errorf("error checking TOML config file: %w", err)
+	}
+
+	return nil
+}
+
+func populateFromCLIContext(cfg *CLIConfig, ctx *cli.Context) {
```
Comment (jakim929) [inline config/cli.go:306]: RC: this function makes it so that if the TOML config is set, the cli flags can't override it. Should update it so that the explicit flags passed to CLI always take precedence. That would align better with the expected behavior of most CLIs (explicit flags override file base config)

We can just remove the checks for the default values (ie. cfg.L1Port == 0)

**Review by jakim929 — CHANGES_REQUESTED** [review]
Thanks for the PR! Mostly looks great Left some comments

1. Explicitly passed flags should always override file based config
2. Not have separate type definitions for CLIConfig
3. Some nit readme updates

**Review by Abuchtela — APPROVED** [review]

File: README.md (hunk 1 above)
Comment (s29papi) [reply README.md]: done. Resolved this.

**Comment by s29papi** [comment]
requesting a review on this

**Review by Abuchtela — CHANGES_REQUESTED** [review]
closes #104

**Review by Abuchtela — CHANGES_REQUESTED** [review]
fixes #295

**Review by jakim929 — APPROVED** [review]
looks good! thanks for addressing all the changes, and sorry for the delay in the review
